lib/src/embedded/isolate_dispatcher.dart

```diff
@@ -18,34 +20,38 @@ import 'util/explicit_close_transformer.dart';
 import 'util/proto_extensions.dart';
 import 'utils.dart';
 
-/// The message sent to a previously-inactive isolate to initiate a new
-/// compilation session.
-///
-/// The [SendPort] is used to establish a dedicated [IsolateChannel] for this
-/// compilation and the [int] is the compilation ID to use on the wire.
-///
-/// We apply the compilation ID in the isolate for efficiency reasons: it allows
-/// us to write the protobuf directly to the same buffer as the wire ID, which
-/// saves a copy for each message.
-typedef _InitialMessage = (SendPort, int);
+/// A persisted mailbox resource lease from the pool that can be transfered over
+/// to new owners.
+class _Lease {
+  /// A mailbox.
+  final Mailbox mailbox;
+
+  /// The compilationId.
+  int id = 0;
+
+  /// The PoolResource.
+  PoolResource resource;
+
+  _Lease(this.mailbox, this.id, this.resource);
+}
```

Comment:
I'm not generally a fan of using mutable state to implicitly pass messages around a program. If we need to track information for a given child isolate, I'd rather keep that logic encapsulated in a class where mutations are done explicitly through method calls.

Reply:
Here is why we need a mutable wrapper to hold reference:

``` dart
  Future<Mailbox> _getMailbox(int compilationId) async {
    // Acquire or create a wrapper object
    var lease = ...;
    ... 
    // Save new compilationId into wrapper object
    lease.id = compilationId;
    ...
    // Because the isolate and channel objects are persisted, callback is only registered once
    channel.stream.listen((message) {
      ... 
      if (message[0] == 1) {
        // Because callback is only registered once, we cannot use compilationId here as it always reference to the first compilationId any persist isolate has handled. 
        // Therefore we need a mutable wrapper that hold a reference to the "current" compilationId 
        _mailboxes.remove(lease.id);
        _inactive.add(lease);
        // Same for the PoolResource object, we need to hold a reference to the "current" one
        lease.resource.release();
      }
      ...
    }
```

I understand having a mutable wrapper is not ideal, however, this is the most performant options among the alternatives:
1. Having extra lookup maps of `Mailbox -> compilationId` and `compilationId -> PoolResource` - map lookup is slower than dereference.
2. Deserialize the varint at the beginning of the outbound message to get the compilationId. This would be already slower than dereference, and would still require an extra map to track `PoolResource`.

Reply:
I added method `void transfer(int, PoolResource)` to the `_Lease` wrapper to hide the mutation of property.

@nex3 I'm open to other suggestion if you have better idea.

Reply:
I've pushed a patch that refactors parts of the lease concept out into a `ReusableIsolate` class that encapsulates all the necessary mutability.